File: 2021/103. Binary Tree Zigzag Level Order Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> ans;
        if (root == NULL) return ans;
        queue<TreeNode*> queue;
        queue.push(root);
        bool leftToRight = true;        
        while (!queue.empty()) {            
            vector<int> tmp;
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                root = queue.front();
                queue.pop();         
                tmp.push_back(root->val);
                if (root->left) queue.push(root->left);
                if (root->right) queue.push(root->right);
            }
            if (!leftToRight) {
                reverse(tmp.begin(), tmp.end());
            }
            ans.push_back(tmp);
            leftToRight = !leftToRight;
        }
        return ans;
    }
};

File: 5-Problem-A-Day/103. Binary Tree Zigzag Level Order Traversal.java
/**
 * Definition for a binary tree node. public class TreeNode { int val; TreeNode left; TreeNode
 * right; TreeNode(int x) { val = x; } }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        LinkedList<List<Integer>> ans = new LinkedList<>();
        if (root == null)
            return ans;
        Queue<TreeNode> nodes = new LinkedList<>();
        Queue<Integer> levels = new LinkedList<>();
        nodes.add(root);
        levels.add(1);
        int reverse = 1;
        while (nodes.size() > 0) {
            TreeNode node = nodes.remove();
            int level = levels.remove();
            if (level > ans.size()) {
                if (reverse == 1 && ans.size() > 0)
                    Collections.reverse(ans.getLast());
                reverse = 1 - reverse;
                ans.add(new LinkedList<>());
            }
            ans.getLast().add(node.val);
            if (node.left != null) {
                nodes.add(node.left);
                levels.add(level + 1);
            }
            if (node.right != null) {
                nodes.add(node.right);
                levels.add(level + 1);
            }
        }
        if (reverse == 1 && ans.size() > 0)
            Collections.reverse(ans.getLast());
        return ans;
    }
}
